pkg/agent/run.go

```diff
@@ -208,6 +208,10 @@ func RunStandalone(ctx context.Context, cfg cmds.Agent) error {
 		close(cfg.AgentReady)
 	}
 
+	if err := tunnel.Setup(ctx, nodeConfig, proxy); err != nil {
+		return err
+	}
```

Comment:
Sorry for digging up old(-ish) changes, but I am investigating [an issue](https://github.com/loft-sh/vcluster/issues/522) with our k3s usage in [vcluster project](https://github.com/loft-sh/vcluster), and I have a question specifically about this change^

Why is the tunnel setup needed when the agent is disabled? Could the etcd-only node have the agent disabled?

Reply:
The tunnel code is responsible for maintaining connections to all the apiserver endpoints; without it the agent is connected to the single endpoint that it connected to when bootstrapping.

Note that disabling the agent really just disables the kubelet and containerd; there is some agent code that is used even when agent is disabled. It really should be called --disable-kubelet-and-containerd.

There are some other bits in the embedded etcd codebase that rely on a node object existing for each etcd cluster member; disabling the agent while using embedded etcd is unlikely to work reliably, especially when adding/removing nodes from the cluster.

Reply:
Thank you for such a quick and detailed answer.
I am not 100% sure about the full impact, but at the very least our k3s logs are spammed with this message:
```
time="2022-06-03T16:15:32.511467095Z" level=error msg="Failed to connect to proxy" error="x509: certificate is valid for 127.0.0.1, 10.100.221.26, 10.100.92.27, not 192.168.62.123"
time="2022-06-03T16:15:32.511496600Z" level=error msg="Remotedialer proxy error" error="x509: certificate is valid for 127.0.0.1, 10.100.221.26, 10.100.92.27, not 192.168.62.123"
```
We are deploying k3s as a pod with [many things disabled](https://github.com/loft-sh/vcluster/blob/main/charts/k3s/values.yaml#L118-L129) (e.g. "agent" - aka kubelet and containerd), and for now it is running as a single replica. The cluster doesn't have any "real" nodes connected to it, and the scheduling is done with our external controller, and the workloads are then run in the same cluster as k3s pod.
The 192.168.62.123 IP mentioned above is IP of the k3s pod, and we write it into default/kubernetes Endpoint resource. When I inspect the certificate in the kube-system/k3s-serving Secret inside of this k3s instance, it does have 192.168.62.123 IP in it, so I am not sure why would remotedialer consider it invalid, am I checking the wrong cert? I think that's where I will focus my attention tomorrow, checking which cert is remotedialer comparing to :thinking:

Reply:
You might see if adding that IP to the --tls-san argument helps? I suspect it's the apiserver cert that's missing that address. I suspect that some of the address autodetection isn't working with the agent disabled. Where are the other two addresses coming from?